Detectors/FIT/FT0/raw/include/FT0Raw/DigitBlockFT0.h

```diff
@@ -0,0 +1,134 @@
+//
+// Artur.Furs
+// afurs@cern.ch
+// TODO:
+//  traites for DataBlocks
+//  check if the EventID filling is correct
+
+#ifndef ALICEO2_FIT_DIGITBLOCKFT0_H_
+#define ALICEO2_FIT_DIGITBLOCKFT0_H_
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <Rtypes.h>
+#include "FT0Raw/DataBlockRaw.h"
+#include <CommonDataFormat/InteractionRecord.h>
+#include "DataFormatsFT0/Digit.h"
+#include "DataFormatsFT0/ChannelData.h"
+#include "DataFormatsFT0/LookUpTable.h"
+
+#include <boost/mpl/vector.hpp>
+#include <boost/mpl/contains.hpp>
+
+#include <gsl/span>
+namespace o2
+{
+namespace ft0
+{
+
+class DigitBlockFT0
+{
+  //Mandatory to fill! Just add class type into corresponding template header.
+  typedef boost::mpl::vector<DataBlockPM, DataBlockTCM, DataBlockTCMext> ProccessBlockTypes;
+  //typedef boost::mpl::vector<DataBlockTCM,DataBlockTCMext> TrgInfoTypes;
+  //
+ public:
+  DigitBlockFT0(o2::InteractionRecord intRec) { mDigit.mIntRecord = intRec; }
+  DigitBlockFT0() = default;
+  DigitBlockFT0(const DigitBlockFT0& other) = default;
+
+  Digit mDigit;
+  std::vector<ChannelData> mVecChannelData;
+  static o2::ft0::LookUpTable sLookupTable;
+  static int sEventID;
+  template <class DataBlockType>
+  //,typename enable_if<boost::mpl::contains<ProccessBlockTypes,DataBlockType>::type::value>::type >  //TODO traits
+  void proccess(DataBlockType& dataBlock, int linkID)
+  {
+    if constexpr (std::is_same<DataBlockType, DataBlockPM>::value) { //Filling data from PM
+      for (int iEventData = 0; iEventData < dataBlock.DataBlockWrapper<RawDataPM>::mNelements; iEventData++) {
+        mVecChannelData.emplace_back(int(sLookupTable.getChannel(linkID, dataBlock.DataBlockWrapper<RawDataPM>::mData[iEventData].channelID)),
+                                     int(dataBlock.DataBlockWrapper<RawDataPM>::mData[iEventData].time),
+                                     int(dataBlock.DataBlockWrapper<RawDataPM>::mData[iEventData].charge),
+                                     int(dataBlock.DataBlockWrapper<RawDataPM>::mData[iEventData].numberADC));
+      }
+    } else if constexpr (std::is_same<DataBlockType, DataBlockTCM>::value) { //Filling data from TCM (normal/extended mode)
+      dataBlock.DataBlockWrapper<RawDataTCM>::mData[0].pushTrgData(mDigit.mTriggers);
+    }
+
+    else if constexpr (std::is_same<DataBlockType, DataBlockTCMext>::value) { //Filling data from TCM, extended mode. Same proccess as for normal mode, for now.
+      dataBlock.DataBlockWrapper<RawDataTCM>::mData[0].pushTrgData(mDigit.mTriggers);
+    }
+  }
+  void popData(std::vector<Digit>& vecDigits, std::vector<ChannelData>& vecChannelData)
+  {
+    //last digit filling
+    mDigit.ref.set(vecChannelData.size(), mVecChannelData.size());
+    mDigit.mEventID = sEventID;
+    //
+    vecDigits.push_back(std::move(mDigit));
+    std::move(mVecChannelData.begin(), mVecChannelData.end(), std::back_inserter(vecChannelData));
```

Comment:
I was thinking about this way of decoding, but the data are processing page by page. In case of constructing ChannelData in-place within final container, it will demand to do decoding process on sequence of raw data blocks with monotonic InteractionRecord increasing, which cannot be achieved in decoding schematic "page by page". So it will require to decode into raw DataBlocks all pages at once, store them in the buffer, calculate first position and number of ChannelData elements for each IR, and then process into the Digits and ChannelData, saving monotonic IR increasing. I guess that it will take some additional operations, buffer(for storing raw data blocks) allocation which needs estimation of the raw data size per TF, etc.

So, if it is really necessary to make strong decoding time optimization, I can change decoding schematics. If I miss something or do some incorrect consideration, please let me know.

Reply:
Another idea is to do parallel partial page decoding, IR by IR, so there will be no need in buffer.

Reply:
ok, let's keep it at the moment as it is, will see if it needs a speed up.